### PR TITLE
Add JAX equivalent of scipy.stats.mode

### PR DESCRIPTION
**Pull request by ylamidon**
Solution for (https://github.com/google/jax/issues/8978)

The previous PR (https://github.com/google/jax/pull/9141) and (https://github.com/google/jax/pull/11475) seemed stale.

This PR is largely based on these previous contributions and will hopefully resolve the remaining issues.

### CODE REVIEW

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:31]: Let's use `x: ArrayLike` (same import as `Array` above)

Also, you check `axis is None` below, should `axis` be `Optional[int]`?

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:27]: Let's use
```python
from typing import NamedTuple
from jax._src.typing import Array
class ModeResult(NamedTuple):
  mode: Array
  count: Array
```

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:52]: 2-space indent

File: jax/scipy/stats/__init__.py
Comment (jakevdp) [inline jax/scipy/stats/__init__.py:36]: `_core` is meant to be an internal-only submodule... instead we should import `mode` directly into this namespace (`jax.scipy.stats.mode` like `scipy.stats.mode`)

File: tests/scipy_stats_test.py
Comment (jakevdp) [inline tests/scipy_stats_test.py:945]: This should be `lsp_stats.mode`

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:29]: Perhaps use `lax_description` to mention that custom `nan_policy` is not supported

File: tests/scipy_stats_test.py
Comment (jakevdp) [inline tests/scipy_stats_test.py:916]: We're gradually moving the tests to use `jtu.sample_product` instead of `parameteriozed.named_parameters` (see #12660). You could optionally do that here if you're feeling up for it! But if not, this looks fine.

**Review by jakevdp — COMMENTED** [review]
Looks great! Just a few minor comments

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:31]: It makes sense
I need to rename the mode input ```x``` to something like ```a``` which is of type ```ArrayLike```
So that I can use ```x = jnp.asarray(a)``` to get ```x``` of type ```Array```
And then I can use methods like ```shape``` or ```ndim``` on ```x```

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:27]: ```NamedTuple``` as the ```count``` method, so I'll use ```counts``` instead. And also ```modes``` instead of ```mode```.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:27]: I think we should keep it as `mode` and `count` to match the numpy API. Those names don't seem to cause issues when Itry them

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:27]: The problem occurs when using mypy

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:27]: OK, in that case let's keep the current approach

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:27]: Correct runtime API is more important than type annotations

File: tests/scipy_stats_test.py
Comment (jakevdp) [inline tests/scipy_stats_test.py:932]: To avoid generating a list of all possible combinations, use something like this:
```python
@jtu.sample_product(
    [dict(shape=shape, axis=axis)
      for shape, axis in (
        ((0,), None),
        ((7,), None),
        ((7,), 0),
        ((47, 8), None),
        ((47, 8), 0),
        ((47, 8), 1),
        ((10, 5, 21), None),
        ((10, 5, 21), 0),
        ((10, 5, 21), 1),
        ((10, 5, 21), 2),
      )
    ],
    dtype=jtu.dtypes.integer + jtu.dtypes.floating,
    contains_nans=[True, False])
```

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:30]: If this causes mypy errors, lets switch it back to how you originally had it defined.

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:38]: Maybe make this `x = jnp.atleast_1d(a)` instead of calling it further down.

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:46]: Is this TODO something you'd like to do as part of this PR, or leave it for later?

**Review by jakevdp — CHANGES_REQUESTED** [review]
Sorry for the delayed review - this is looking great! Just a couple more minor comments.

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:46]: If it's not required, maybe we can leave it for later

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:37]: `atleast_1d` calls `asarray` on the input, so no need to call both here. `x = jnp.atleast_1d(a)` should be sufficient.

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:39]: Use a list rather than a set here, because if a user passes a non-hashable value for `nan_policy` this membership check will raise an error.

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:45]: Sounds good

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:56]: This indentation still needs to be fixed

File: jax/scipy/stats/__init__.py
Comment (jakevdp) [inline jax/scipy/stats/__init__.py:36]:
```suggestion
from jax.scipy.stats import mode
```
`jax.scipy.stats.mode` is the public import path for this function, so we should use it in the tests.

Comment (ylamidon) [reply jax/scipy/stats/__init__.py:36]: I'm not sure to understand.
In the file  ```jax/scipy/stats/__init__.py```, the current import seems ok.
In the test file (```scipy_stats_test.py```), we do:
```
from jax.scipy import stats as lsp_stats
lax_fun = partial(lsp_stats.mode, axis=axis)
```
Which seems ok.

Comment (jakevdp) [reply jax/scipy/stats/__init__.py:36]: Sorry, you're right. I thought this was the test file

**Review by jakevdp — COMMENTED** [review]
Looks great, thanks for sticking with this! Last two things before we can merge:

- the comment above about indentation: all indentations should be two spaces
- once that is fixed, please squash all changes into a single commit (see https://jax.readthedocs.io/en/latest/contributing.html#single-change-commits-and-pull-requests)

Thanks!

**Review by jakevdp — APPROVED** [review]
Looks great! Thanks for the contribution

**Comment by jakevdp** [comment]
It looks like we're hitting a `FutureWarning` in the scipy version of the function. I'd suggest we add a `keepdims` argument that defaults to `False` (i.e. the future default behavior for the `scipy` version) and expand the test cases to cover that. What do you think?

**Comment by ylamidon** [comment]
I tried to add ```keepdims``` to the mode function with:
```
  if keepdims:
    return ModeResult(vals.reshape(output_shape), counts.reshape(output_shape))
  else:
    return ModeResult(jnp.squeeze(vals.reshape(output_shape), axis=axis), jnp.squeeze(counts.reshape(output_shape), axis=axis))
```
But this produces in a ```ConcretizationTypeError```

**Comment by jakevdp** [comment]
Did you somehow convert `output_shape` or one of its elements to a jax array? You should make sure it is a tuple/list of integers.

**Comment by jakevdp** [comment]
Seems like it may be more straightforward to change line 67, adjusting the `output_shape` depending on the value of `keepdims`.

**Comment by jakevdp** [comment]
Oh, wait, I bet the issue is probably that you passed `keepdims` as a dynamic argument. It would have to be a static argument like `axis`. But the note about `out_shape` still holds.

**Comment by ylamidon** [comment]
Yes, I forgot to add ```keepdims``` to ```static_argnames```

**Comment by ylamidon** [comment]
When ```shape = (0,), axis=None``` and ```keepdims``` is ```True```  or ```False```,  numpy produces a RuntimeWarning: Mean of empty slice.
So I will probably have to remove the test with ```shape = (0,), axis=None```.

**Comment by jakevdp** [comment]
Don't remove the test, instead filter the expected warnings (search the codebase for `jtu.ignore_warning` to see examples of this)

**Comment by ylamidon** [comment]
```jtu.ignore_warning works``` fine.
For ```x = np.zeros((0,))```
```scipy.stats.mode(x, axis=None, keepdims=True)``` returns ```ModeResult(mode=array([nan]), count=array([nan]))```.
And
```jax.scipy.stats.mode(x, axis=None, keepdims=True)``` returns ```ModeResult(mode=DeviceArray([], dtype=float32), count=DeviceArray([], dtype=int32))```

**Comment by jakevdp** [comment]
It looks like we'll have to handle zero-size arrays differently to match the semantics of `scipy.stats.mode`; for example:
```python
>>> scipy.stats.mode(np.ones((0, 2, 3)), axis=2, keepdims=True).mode.shape                                                                                                                            
(0, 2, 1)
```

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:60]: `jnp.nan` is a python scalar `NaN`... for consistency we should return a jax array in all cases. We should use `jnp.full` in both cases... maybe compute the output_shape within the `if` statement and only have a single return statement? Maybe even one `return` statement for the entire `if x.size == 0` block? The only thing that's changing is the output shape, yes?

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: Also this is a bit in the weeds, but `jnp.full(shape, jnp.nan)` returns a weakly-typed array because `jnp.nan` is a Python scalar. For consistency with the non-empty return value we should use
```python
jnp.full(shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))
```

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [inline jax/_src/scipy/stats/_core.py:82]: Does this result in the correct shape for `axis=None` with `keepdims=True`?

File: jax/_src/scipy/stats/_core.py
Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: Another thought... this may all be cleaner if you handle the empty case in `_mode_helper` instead.

**Review by jakevdp — CHANGES_REQUESTED** [review]

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:60]: ```scipy.stats.mode(np.zeros((0,)), axis=None, keepdims=False)``` and ```scipy.stats.mode(np.zeros((0,)), axis=0, keepdims=False)``` both return ```ModeResult(mode=nan, count=nan)```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: But they don't return a Python scalar NaN, they return a numpy.float64 NaN (i.e. a numpy scalar). JAX's equivalent of a numpy scalar is a zero-dimensional array.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: This is important because, for example, someone might write code using something like `result.mode.shape`, and if we return a Python scalar NaN in some cases, that would result in an AttributeError. For that reason we generally try to keep return types consistent for different shape inputs. Does that make sense?

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:82]: Yes, I think this part of the code is fine for axis=None with keepdims=True.

File: jax/_src/scipy/stats/_core.py
Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:60]: Yes it makes sense.
So instead of returning ```ModeResult(jnp.nan, jnp.nan)``` in this case, I can return  
```ModeResult(jnp.full((), jnp.nan), jnp.full((), jnp.nan))``` 
?

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: Yes, basically, except we should specify the dtype: `jnp.full(shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))`

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py:60]: To make the code a bit cleaner, I can refactor it like this:
```python
  def _mode_result_nan(output_shape):
    return ModeResult(
      jnp.full(output_shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)),
      jnp.full(output_shape, jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)))

  def _get_output_shape(shape):
    output_shape = list(shape)
    if keepdims:
      output_shape[axis] = 1
    else:
      del output_shape[axis]
    return output_shape

  if x.size == 0:
    if axis is None:
      output_shape = [1 for _ in range(x.ndim)] if keepdims else ()
      return _mode_result_nan(output_shape)
    else:
      return _mode_result_nan(_get_output_shape(x.shape))
```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: I'd avoid that kind of extra indirection in the form of one-line helper functions. I was just saying that you might define `output_shape` within the `if`/`else` blocks and then have a single return statement at the end. But if you prefer multiple return statements, that's fine too.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py:60]: I think the best approach would be to handle the empty case within the vmapped helper, but we've already been going a long time here and I don't want to make you refactor what you have too much!

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: So i added a helper function and pushed the changes.   
I'm not sure that's exactly what you wanted for the helper function.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: Sorry, I think I was unclear. I don't think we need all these extra helper functions for this algorithm: it's too much indirection and makes the code much harder to follow.

When I said "handle the empty case within the vmapped helper", I meant something like this:
```python
  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
    """Helper function to return mode and count of a given array."""
    if x.size == 0:
      return jnp.float32(jnp.nan), jnp.float32(jnp.nan)
    else:
      vals, counts = jnp.unique(x, return_counts=True, size=x.size)
      return vals[jnp.argmax(counts)], counts.max()
```
Then I think the zero-size case would be handled naturally within the `vmap` where this function is applied.

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Yes, but then I have error in the following cases:
```python
shape=(0,), axis=None, keepdims=False
shape=(0,), axis=None, keepdims=True
shape=(0,), axis=0, keepdims=False
shape=(0,), axis=0, keepdims=True
shape=(0, 2, 3), axis=None, keepdims=False
shape=(0, 2, 3), axis=None, keepdims=True
shape=(0, 2, 3), axis=0, keepdims=False
shape=(0, 2, 3), axis=0, keepdims=True
```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: Without seeing the error, I suspect it's coming from `x = x.reshape(x.shape[0], -1)` which is incompatible with zero-sized arrays. you can do `x = x.reshape(x.shape[0], prod(x.shape[1:]))` instead and that will work for arrays of any size

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Yes, thanks. But I also have errors like:
```python
ValueError: vmap was requested to map its argument along axis 1, which implies that its rank should be at least 2, but is only 1 (its shape is (0,))
```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: Hard to guess without seeing the code. It sound like you removed the `reshape` line if your array has only one dimension when you pass it to the vmapped function

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Sorry, my mistake. There is another error:
```
jax._src.traceback_util.UnfilteredStackTrace: UnboundLocalError: local variable 'vals' referenced before assignment
```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: Something like this should work:
```python
def mode(a: ArrayLike, axis: Optional[int] = 0, nan_policy: str = "propagate", keepdims: bool = False) -> ModeResult:
  _check_arraylike("mode", a)
  x = jnp.atleast_1d(a)

  if nan_policy not in ["propagate", "omit", "raise"]:
    raise ValueError(
      f"Illegal nan_policy value {nan_policy!r}; expected one of "
      "{'propoagate', 'omit', 'raise'}"
    )
  if nan_policy == "omit":
    # TODO: return answer without nans included.
    raise NotImplementedError(
      f"Logic for `nan_policy` of {nan_policy} is not implemented"
    )
  if nan_policy == "raise":
    raise NotImplementedError(
      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
      "Please check if nans exist in `x` outside of the `mode` function."
    )

  input_shape = x.shape
  if keepdims:
    if axis is None:
      output_shape = tuple(1 for i in input_shape)
    else:
      output_shape = tuple(1 if i == axis else s for i, s in enumerate(input_shape))
  else:
    if axis is None:
      output_shape = ()
    else:
      output_shape = tuple(s for i, s in enumerate(input_shape) if i != axis)

  if axis is None:
    axis = 0
    x = x.ravel()

  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
    """Helper function to return mode and count of a given array."""
    if x.size == 0:
      return jnp.array(jnp.nan, dtype=x.dtype), jnp.float32(jnp.nan)
    else:
      vals, counts = jnp.unique(x, return_counts=True, size=x.size)
      return vals[jnp.argmax(counts)], counts.max()

  axis = canonicalize_axis(axis, x.ndim)
  x = jnp.moveaxis(x, axis, 0)
  x = x.reshape(x.shape[0], prod(x.shape[1:]))
  vals, counts = vmap(_mode_helper, in_axes=1)(x)
  return ModeResult(vals.reshape(output_shape), counts.reshape(output_shape))
```
It won't match scipy for `keepdims=True` and `axis=None`, but scipy is wrong in that case.

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: I just realized we can't return `jnp.array(jnp.nan, dtype=x.dtype)` because `x` may be integer typed. I'm not sure what's best in that case.

If you want to go back to what you had yesterday (before the helper function refactor) I'd be fine with that too. Your call

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: But we should make sure we *don't* match scipy for `keepdims=True` and `axis=None`. The return shape in that case should be `(1,) * len(input_shape)`

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Just to clarify:
```python
scipy.stats.mode(np.ones((0, 3, 2)), axis=None, keepdims=True).mode.shape
```
returns ```(1, 1, 1)```.
And
```python
scipy.stats.mode(np.ones((4, 3, 2)), axis=None, keepdims=True).mode.shape
```
returns ```(1,)```.

So, ```jax.scipy.stats.mode``` should replicate ```scipy.stats.mode``` in both cases ?

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: Sorry, I mixed that up. I think it should return `(1, 1, 1)` in both cases. From https://github.com/scipy/scipy/issues/17214#issuecomment-1276464112 it's my understanding that this is the intended API.

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Ok. 
But then the tests where x is not empty, ```ndim>=2 , axis=None, keepdims=True``` will fail because the shape returned by ```jax.scipy.stats.mode``` is not the same as the shaped returned by ```scipy.stats.mode```.
Is there a way to deactivate the test on the shape in these cases ?

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Another thing. Shouldn't ```_mode_helper``` return 
```python
return jnp.array(jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_)), jnp.array(jnp.nan, dtype=dtypes.canonicalize_dtype(jnp.float_))
```
when ```x.size== 0```

Comment (jakevdp) [reply jax/_src/scipy/stats/_core.py]: > Is there a way to deactivate the test on the shape in these cases ?

I'd probably approach this by writing a wrapper around the scipy function that detects this case and returns the expected result.

> Shouldn't _mode_helper return...

Yes, you're right

Comment (ylamidon) [reply jax/_src/scipy/stats/_core.py]: Ok. I made the changes and pushed the result

**Review by jakevdp — APPROVED** [review]
Looks great! Thanks for sticking with this & sorry for so many review comments 😀 I'm going to trigger tests and hopefully this will all pass 🤞

**Comment by jakevdp** [comment]
So, one last issue: `keepdims` wasn't added to scipy until version 1.9.0, and some of our test invocations still use this older version of scipy.

What we'll want to do is compute `scipy_version = tuple(map(int, scipy.__version__.split('.')[:3]))`, and then if `scipy_version < (1, 9, 0)` avoid passing `keepdims` to `scipy.stats.mode` (and probably just test `keepdims = True` for those versions, which is the default behavior for older scipy versions)

**Comment by ylamidon** [comment]
I did the dev to avoid passing ```keepdims``` to ```scipy.stats.mode``` for ```scipy_version < (1, 9, 0)```.  

There is a small problem with the 1.7.3 version of scipy that I used for testing.  
* With the scipy version 1.9.2, ```scipy.stats.mode(a, keepdims=True)``` return an array of size 1 for an input array a of size 0
* With the scipy version 1.7.3, ```scipy.stats.mode(a) ``` return an array of size 0 for an input array a of size 0  

The problem is that I cannot reshape an array of size 0 to a shape of ```(1,) * ndim```.

So maybe I can use the ```scipy_mode_wrapper``` to return a pair of ```np.full(output_shape, np.nan, dtype=dtypes.canonicalize_dtype(jax.numpy.float_)``` instead of calling ```scipy.stats.mode ``` in this case ?

**Comment by jakevdp** [comment]
Sure, that sounds like a good approach.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Looks like there's still an error on scipy 1.5:
```
=================================== FAILURES ===================================
______________________ LaxBackedScipyStatsTests.testMode4 ______________________
[gw6] linux -- Python 3.7.13 /usr/bin/python3.7
tests/scipy_stats_test.py:949: in testMode
    tol=tol)
jax/_src/test_util.py:936: in _CheckAgainstNumpy
    canonicalize_dtypes=canonicalize_dtypes)
jax/_src/test_util.py:848: in assertAllClose
    err_msg=err_msg)
jax/_src/test_util.py:856: in assertAllClose
    err_msg=err_msg)
jax/_src/test_util.py:820: in assertArraysAllClose
    _assert_numpy_allclose(x, y, atol=atol, rtol=rtol, err_msg=err_msg)
jax/_src/public_test_util.py:96: in _assert_numpy_allclose
    np.testing.assert_allclose(a, b, **kw, err_msg=err_msg)
E   AssertionError:
E   Not equal to tolerance rtol=0.0002, atol=0.0002
E
E   x and y nan location mismatch:
E    x: array([[[-8.991903]]], dtype=float32)
E    y: array([[[nan]]], dtype=float32)
```

**Comment by ylamidon** [comment]
Yes. Actually, its fine with scipy 1.5.4 and a new version of numpy. But there is an issue with scipy 1.5.4 and numpy 1.20.3.

**Comment by jakevdp** [comment]
We might be able to bump the minimum scipy version to 1.7 - give me a bit to figure out if that's possible

**Comment by jakevdp** [comment]
I think it's probably fine to skip the tests for scipy versions older than 1.6. We may not bump the minimum scipy right away, but it will happen soon-ish.

**Comment by ylamidon** [comment]
I noticed that the tests are ok with numpy 1.21.0 and scipy 1.5.4 but not ok with numpy 1.20.0 and scipy 1.5.4.
The problem occurs when ```contains_nans = True```.
So maybe I can skip the tests with ```contains_nans = True``` when numpy version < 1.21.0 ?

**Comment by jakevdp** [comment]
Sure, that sounds like a good plan.

**Review by jakevdp — APPROVED** [review]

**Comment by jakevdp** [comment]
Whew, nice work on this! It ended up being quite a bit more complicated than anticipated with all the numpy/scipy version effects, thanks so much for sticking with it through the whole process!

**Comment by ylamidon** [comment]
Thank you. Yes, it was a bit more complicated than expected. Thanks so much for all the help and valuable feedback!

**Pull request merged**
